File: backend/laravel/app/Models/ComunidadeMembro.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\Pivot;

#[Fillable(['comunidade_id', 'usuario_id', 'papel', 'adicionado_por_id'])]
class ComunidadeMembro extends Pivot
{
    public const ORGANIZADOR = 'organizador';

    public const MEMBRO = 'membro';

    public $incrementing = true;

    public $timestamps = false;

    protected $table = 'comunidade_membros';

    protected $attributes = [
        'papel' => self::MEMBRO,
    ];

    protected static function booted(): void
    {
        static::creating(function (ComunidadeMembro $vinculo) {
            if (! $vinculo->adicionado_em) {
                $vinculo->adicionado_em = now();
            }
        });
    }

    protected function casts(): array
    {
        return [
            'adicionado_em' => 'datetime',
        ];
    }

    public function comunidade(): BelongsTo
    {
        return $this->belongsTo(Comunidade::class);
    }

    public function usuario(): BelongsTo
    {
        return $this->belongsTo(User::class, 'usuario_id');
    }

    public function adicionadoPor(): BelongsTo
    {
        return $this->belongsTo(User::class, 'adicionado_por_id');
    }
}

File: backend/rails/app/models/comunidade_membro.rb
# Vínculo de um usuário com uma comunidade e seu papel nela (organizador
# ou membro). Espelha core/models.py:ComunidadeMembro (Django) /
# app/Models/ComunidadeMembro.php (Laravel).
class ComunidadeMembro < ApplicationRecord
  ORGANIZADOR = "organizador"
  MEMBRO = "membro"
  PAPEIS = [ORGANIZADOR, MEMBRO].freeze

  belongs_to :comunidade, inverse_of: :comunidade_membros
  belongs_to :usuario, class_name: "User", inverse_of: :comunidade_membros
  belongs_to :adicionado_por, class_name: "User", optional: true

  before_validation { self.adicionado_em = Time.current if adicionado_em.blank? }

  validates :papel, presence: true, inclusion: { in: PAPEIS }
  validates :usuario_id, uniqueness: { scope: :comunidade_id }
end
